### PR TITLE
[IMP] web, *: allow user's to quick login

### PR DESCRIPTION
**Pull request by res-odoo**
*: auth_oauth, auth_passkey, auth_signup, auth_totp

When several users share the same device, the login/logout process can
be boring, so we added a quick login feature.

For a security propose it doesn't store the session or the password of
the user. No data is sent to the server, everything is stored inside
the browser localStorage[1] . It only stores the login of the user and
the avatar of the user for the last 5 users connected to the device.

We also lightly redesign the login screen

task-3953707

[1]: https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage

### CODE REVIEW

File: addons/web/static/tests/tours/user_switch_tour.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
+/** @odoo-module **/
```
Comment (rfr-odoo) [inline addons/web/static/tests/tours/user_switch_tour.js:1]: Still needed ?

Comment (res-odoo) [reply addons/web/static/tests/tours/user_switch_tour.js:1]: Nop, well catch

File: addons/web/static/tests/tours/user_switch_tour.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                    users[0].parentElement.querySelector(".fs-5").innerText.trim(),
+                    "Mitchell Admin"
+                );
+                assertEqual(
+                    users[1].parentElement.querySelector(".fs-5").innerText.trim(),
+                    "Marc Demo"
+                );
+                users[1].click();
+            },
+        },
```
Comment (rfr-odoo) [inline addons/web/static/tests/tours/user_switch_tour.js:86]: Check can be done with step no need to do all complex `run` function
e.g.:
```suggestion
        {
            content: "Check if there is Mitchell Admin in user list selection",
            trigger: ".o_user_switch .list-group-item:nth-child(1):contains('Mitchell Admin')",
        },
        {
            content: "Check if there is Marc Demo in user list selection",
            trigger: ".o_user_switch .list-group-item:nth-child(2):contains('Marc Demo')",
        },
        {
            content: "Choice demo",
            trigger: ".o_user_switch .list-group-item:contains('Marc Demo')",
            run: "click",
            },
        },
```

File: addons/web/static/tests/tours/user_switch_tour.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            trigger: ".o_user_switch",
+            run() {
+                const users = document.querySelectorAll(".o_user_avatar");
+                assertEqual(users.length, 1);
+                assertEqual(
+                    users[0].parentElement.querySelector(".fs-5").innerText.trim(),
+                    "Marc Demo"
+                );
+                users[0].click();
+            },
```
Comment (rfr-odoo) [inline addons/web/static/tests/tours/user_switch_tour.js:42]: Same as previous comment

File: addons/web/static/src/core/user_switch/user_switch.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import { registry } from "@web/core/registry";
+import { lastUsers, registerUsers } from "@web/core/user";
+import { imageUrl } from "@web/core/utils/urls";
+
+export class UserSwitch extends Component {
+    static template = "web.login_user_switch";
+    static props = {};
+
+    setup() {
+        super.setup();
```
Comment (rfr-odoo) [inline addons/web/static/src/core/user_switch/user_switch.js:11]: No needed

File: addons/web/static/src/core/user_switch/user_switch.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            users,
+            displayUserChoice: users.length,
+        });
+        this.form = document.querySelector("form.oe_login_form");
+        if (users.length) {
+            this._maskInputForm();
+        } else {
+            this._unmaskInputForm();
+            this.form.querySelector("input#login").focus();
+        }
```
Comment (rfr-odoo) [inline addons/web/static/src/core/user_switch/user_switch.js:22]: Same logic at line 35 to 40 maybe refactor this into a function with a bool as parameter

File: addons/web/static/src/core/user_switch/user_switch.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            this.fillForm();
+        }
+    }
+
+    fillForm(login = "") {
+        this.form.querySelector("input#login").value = login;
+        this.form.querySelector("input#password").value = "";
+        this.toggleFormDisplay();
+        const input = login.length ? "input#password" : "input#login";
+        this.form.querySelector(input).focus();
```
Comment (adr-odoo) [inline addons/web/static/src/core/user_switch/user_switch.js:59]: unwanted focus?

File: addons/web/static/src/core/user_switch/user_switch.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        const users = lastUsers();
+        this.state = useState({
+            users,
+            displayUserChoice: users.length,
+        });
+        this.form = document.querySelector("form.oe_login_form");
+        this._toggleFormDisplay(users.length);
+    }
+
+    _maskInputForm() {
```
Comment (adr-odoo) [inline addons/web/static/src/core/user_switch/user_switch.js:20]: only used once

File: addons/web/static/src/core/user.js [hunk 8]
```diff
@@ -126,3 +127,33 @@ export function _makeUser(session) {
...
+    return lastConnectedUsers ? JSON.parse(lastConnectedUsers) : [];
+};
+
+export const registerUsers = (users) => {
+    browser.localStorage.setItem(LAST_CONNECTED_USER_KEY, JSON.stringify(users));
+};
+
+if (user.login && user.login !== "__system__") {
+    const users = lastUsers();
+    const lastedUsers = [
```
Comment (adr-odoo) [inline addons/web/static/src/core/user.js:144]: lastUsers?

File: addons/web/static/src/core/user.js [hunk 9]
```diff
@@ -126,3 +127,33 @@ export function _makeUser(session) {
...
+        {
+            login: user.login,
+            name: user.name,
+            partnerId: user.partnerId,
+            partnerWriteDate: user.writeDate,
+            userId: user.userId,
+        },
+    ];
+    for (let i = 0; i < Math.min(5, users.length); i++) {
```
Comment (adr-odoo) [inline addons/web/static/src/core/user.js:153]: to move in the lastUsers()

File: addons/web/static/src/core/user_switch/user_switch.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        this.form.classList.add("d-none");
+    }
+
+    _toggleFormDisplay(displayUserChoice) {
+        if (displayUserChoice) {
+            this._maskInputForm();
```
Comment (adr-odoo) [inline addons/web/static/src/core/user_switch/user_switch.js:26]: Use .toogle("d-none", displayUserChoice) ?

File: addons/web/static/src/webclient/user_menu/user_menu_items.js [hunk 11]
```diff
@@ -124,7 +124,7 @@ function logOutItem(env) {
     return {
         type: "item",
         id: "logout",
-        description: _t("Log out"),
+        description: _t("Log out / Switch User"),
```
Comment (adr-odoo) [inline addons/web/static/src/webclient/user_menu/user_menu_items.js:127]: to check

**Review by adr-odoo — APPROVED** [review]
@robodoo r+

**Review by adr-odoo — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**